File: tests/test_doom_ai.py
import unittest

from doom_actors import MF_SOLID, Actor, MONSTER_TYPES, Player
from doom_ai import (
    check_melee_range, check_missile_range, chase, face_target,
    has_line_of_sight, hitscan_attack, look_for_players, melee_attack,
    pain_and_death,
)
from doom_map import LineDef, Sector, SideDef, Vertex


def _solid_wall(x1, y1, x2, y2):
    sector = Sector(0, 0, "F", "C", 160, 0, 0)
    side = SideDef(0, 0, "-", "-", "WALL", sector)
    return LineDef(Vertex(x1, y1), Vertex(x2, y2), flags=0, special=0, tag=0,
                    frontSide=side, backSide=None)


class LineOfSightTest(unittest.TestCase):
    def test_open_space_has_sight(self):
        self.assertTrue(has_line_of_sight(0, 0, 100, 0, []))

    def test_wall_between_blocks_sight(self):
        wall = _solid_wall(50, -50, 50, 50)
        self.assertFalse(has_line_of_sight(0, 0, 100, 0, [wall]))

    def test_wall_off_to_the_side_does_not_block(self):
        wall = _solid_wall(50, 100, 50, 150)
        self.assertTrue(has_line_of_sight(0, 0, 100, 0, [wall]))


class LookForPlayersTest(unittest.TestCase):
    """ai.md의 '## Feature: 대기 중 플레이어 탐지 (A_Look)' 검증."""

    def test_visible_player_becomes_target_and_triggers_see_state(self):
        zombie = Actor.Spawn(MONSTER_TYPES["Zombieman"], 0, 0)
        playerActor = Actor.Spawn(MONSTER_TYPES["Zombieman"], 100, 0)
        player = Player(playerActor)
        found = look_for_players(zombie, [player], lineDefs=[])
        self.assertTrue(found)
        self.assertIs(zombie.target, playerActor)
        self.assertIs(zombie.state, zombie.type.seeState)

    def test_player_behind_wall_is_not_found(self):
        zombie = Actor.Spawn(MONSTER_TYPES["Zombieman"], 0, 0)
        playerActor = Actor.Spawn(MONSTER_TYPES["Zombieman"], 100, 0)
        player = Player(playerActor)
        wall = _solid_wall(50, -50, 50, 50)
        found = look_for_players(zombie, [player], lineDefs=[wall])
        self.assertFalse(found)
        self.assertIsNone(zombie.target)


class MeleeRangeTest(unittest.TestCase):
    def test_close_target_is_in_melee_range(self):
        demon = Actor.Spawn(MONSTER_TYPES["Demon"], 0, 0)
        demon.target = Actor.Spawn(MONSTER_TYPES["Zombieman"], 10, 0)
        self.assertTrue(check_melee_range(demon))

    def test_far_target_is_not_in_melee_range(self):
        demon = Actor.Spawn(MONSTER_TYPES["Demon"], 0, 0)
        demon.target = Actor.Spawn(MONSTER_TYPES["Zombieman"], 500, 0)
        self.assertFalse(check_melee_range(demon))

    def test_monster_without_melee_state_never_melees(self):
        zombie = Actor.Spawn(MONSTER_TYPES["Zombieman"], 0, 0)  # meleeState=None.
        zombie.target = Actor.Spawn(MONSTER_TYPES["Zombieman"], 5, 0)
        self.assertFalse(check_melee_range(zombie))


class FaceTargetTest(unittest.TestCase):
    def test_faces_directly_at_target(self):
        a = Actor.Spawn(MONSTER_TYPES["Imp"], 0, 0)
        a.target = Actor.Spawn(MONSTER_TYPES["Zombieman"], 0, 100)
        face_target(a)
        self.assertAlmostEqual(a.angle, 90.0)


class MeleeAttackTest(unittest.TestCase):
    """ai.md의 '## Feature: 근접 공격 판정' 검증."""

    def test_deals_damage_when_still_in_range(self):
        demon = Actor.Spawn(MONSTER_TYPES["Demon"], 0, 0)
        victim = Actor.Spawn(MONSTER_TYPES["Zombieman"], 10, 0)
        demon.target = victim
        dealt = []

        def deal_damage(target, damage, inflictor):
            dealt.append((target, damage, inflictor))
            target.health -= damage

        melee_attack(demon, lambda rng: 24, rng=lambda: 0, deal_damage=deal_damage)
        self.assertEqual(len(dealt), 1)
        self.assertIs(dealt[0][0], victim)
        self.assertEqual(dealt[0][1], 24)

    def test_no_damage_if_target_moved_away(self):
        demon = Actor.Spawn(MONSTER_TYPES["Demon"], 0, 0)
        victim = Actor.Spawn(MONSTER_TYPES["Zombieman"], 500, 0)
        demon.target = victim
        dealt = []
        melee_attack(demon, lambda rng: 24, rng=lambda: 0,
                      deal_damage=lambda t, d, i: dealt.append(t))
        self.assertEqual(dealt, [])


class PainAndDeathTest(unittest.TestCase):
    """ai.md의 '## Feature: 고통과 죽음 (A_Pain / A_Fall 계열)' 검증."""

    def test_health_above_zero_may_enter_pain_state(self):
        zombie = Actor.Spawn(MONSTER_TYPES["Zombieman"], 0, 0)
        zombie.health = 10
        result = pain_and_death(zombie, rng=lambda: 0.0)  # 항상 고통 확률을 넘긴다.
        self.assertEqual(result, "pain")
        self.assertIs(zombie.state, zombie.type.painState)

    def test_health_zero_enters_death_state_and_loses_solid(self):
        zombie = Actor.Spawn(MONSTER_TYPES["Zombieman"], 0, 0)
        zombie.health = 0
        result = pain_and_death(zombie, rng=lambda: 0.0)
        self.assertEqual(result, "death")
        self.assertIs(zombie.state, zombie.type.deathState)
        self.assertFalse(zombie.flags & MF_SOLID)


if __name__ == "__main__":
    unittest.main()

File: tests/cpp/test_doom_ai.cpp
// examples/doom/ai.md 검증 (C++17).
#include "doom_ai.hpp"
#include "minitest.hpp"

using namespace Doom;
using namespace Doom::Map;

static LineDef* makeSolidWall(std::vector<std::unique_ptr<Vertex>>& vstore,
                               std::vector<std::unique_ptr<Sector>>& sstore,
                               std::vector<std::unique_ptr<SideDef>>& sdstore,
                               std::vector<std::unique_ptr<LineDef>>& lstore,
                               float x1, float y1, float x2, float y2) {
    vstore.push_back(std::make_unique<Vertex>(Vertex{x1, y1}));
    vstore.push_back(std::make_unique<Vertex>(Vertex{x2, y2}));
    sstore.push_back(std::make_unique<Sector>(Sector{0, 0, "F", "C", 160, 0, 0}));
    sdstore.push_back(std::make_unique<SideDef>(SideDef{0, 0, "-", "-", "WALL", sstore.back().get()}));
    lstore.push_back(std::make_unique<LineDef>(
        LineDef{vstore[vstore.size() - 2].get(), vstore.back().get(), 0, 0, 0, sdstore.back().get(), nullptr}));
    return lstore.back().get();
}

TEST(los_open_space_has_sight) {
    std::vector<LineDef*> none;
    CHECK(hasLineOfSight(0, 0, 100, 0, none));
}

TEST(los_wall_between_blocks_sight) {
    std::vector<std::unique_ptr<Vertex>> vs; std::vector<std::unique_ptr<Sector>> ss;
    std::vector<std::unique_ptr<SideDef>> sds; std::vector<std::unique_ptr<LineDef>> ls;
    auto* wall = makeSolidWall(vs, ss, sds, ls, 50, -50, 50, 50);
    std::vector<LineDef*> lines = {wall};
    CHECK(!hasLineOfSight(0, 0, 100, 0, lines));
}

TEST(los_wall_off_to_the_side_does_not_block) {
    std::vector<std::unique_ptr<Vertex>> vs; std::vector<std::unique_ptr<Sector>> ss;
    std::vector<std::unique_ptr<SideDef>> sds; std::vector<std::unique_ptr<LineDef>> ls;
    auto* wall = makeSolidWall(vs, ss, sds, ls, 50, 100, 50, 150);
    std::vector<LineDef*> lines = {wall};
    CHECK(hasLineOfSight(0, 0, 100, 0, lines));
}

TEST(look_for_players_visible_becomes_target) {
    auto zombie = Actor::Spawn(monsterTypes().at("Zombieman"), 0, 0);
    auto playerActor = Actor::Spawn(monsterTypes().at("Zombieman"), 100, 0);
    Player player(playerActor);
    std::vector<LineDef*> none;
    std::vector<Player*> players = {&player};
    bool found = lookForPlayers(zombie.get(), players, none);
    CHECK(found);
    CHECK(zombie->target == playerActor.get());
    CHECK(zombie->state == zombie->type->seeState);
}

TEST(look_for_players_blocked_by_wall_not_found) {
    std::vector<std::unique_ptr<Vertex>> vs; std::vector<std::unique_ptr<Sector>> ss;
    std::vector<std::unique_ptr<SideDef>> sds; std::vector<std::unique_ptr<LineDef>> ls;
    auto* wall = makeSolidWall(vs, ss, sds, ls, 50, -50, 50, 50);
    auto zombie = Actor::Spawn(monsterTypes().at("Zombieman"), 0, 0);
    auto playerActor = Actor::Spawn(monsterTypes().at("Zombieman"), 100, 0);
    Player player(playerActor);
    std::vector<LineDef*> lines = {wall};
    std::vector<Player*> players = {&player};
    CHECK(!lookForPlayers(zombie.get(), players, lines));
    CHECK(zombie->target == nullptr);
}

TEST(melee_range_close_target) {
    auto demon = Actor::Spawn(monsterTypes().at("Demon"), 0, 0);
    auto victim = Actor::Spawn(monsterTypes().at("Zombieman"), 10, 0);
    demon->target = victim.get();
    CHECK(checkMeleeRange(demon.get()));
}

TEST(melee_range_far_target) {
    auto demon = Actor::Spawn(monsterTypes().at("Demon"), 0, 0);
    auto victim = Actor::Spawn(monsterTypes().at("Zombieman"), 500, 0);
    demon->target = victim.get();
    CHECK(!checkMeleeRange(demon.get()));
}

TEST(face_target_points_directly_at_target) {
    auto a = Actor::Spawn(monsterTypes().at("Imp"), 0, 0);
    auto t = Actor::Spawn(monsterTypes().at("Zombieman"), 0, 100);
    a->target = t.get();
    faceTarget(a.get());
    CHECK(std::abs(a->angle - 90.0f) < 0.001f);
}

TEST(melee_attack_deals_damage_when_in_range) {
    auto demon = Actor::Spawn(monsterTypes().at("Demon"), 0, 0);
    auto victim = Actor::Spawn(monsterTypes().at("Zombieman"), 10, 0);
    demon->target = victim.get();
    int dealtDamage = -1;
    meleeAttack(demon.get(), [](std::function<int()>) { return 24; }, [] { return 0; },
                [&](Actor* t, int d, Actor*) { dealtDamage = d; t->health -= d; });
    CHECK_EQ(dealtDamage, 24);
}

TEST(pain_and_death_high_health_may_pain) {
    auto zombie = Actor::Spawn(monsterTypes().at("Zombieman"), 0, 0);
    zombie->health = 10;
    auto result = painAndDeath(zombie.get(), [] { return 0.0; });
    CHECK_EQ(result, std::string("pain"));
    CHECK(zombie->state == zombie->type->painState);
}

TEST(pain_and_death_zero_health_dies_and_loses_solid) {
    auto zombie = Actor::Spawn(monsterTypes().at("Zombieman"), 0, 0);
    zombie->health = 0;
    auto result = painAndDeath(zombie.get(), [] { return 0.0; });
    CHECK_EQ(result, std::string("death"));
    CHECK(zombie->state == zombie->type->deathState);
    CHECK((zombie->flags & MF_SOLID) == 0);
}

TEST_MAIN()
